esp32/modfreedomgfx.c

```diff
@@ -0,0 +1,273 @@
+  mp_uint_t len;
+  const unsigned char *data = (const unsigned char*) mp_obj_str_get_data(args[2], &len);
+  int x0 = mp_obj_get_int(args[0]);
+  int y0 = mp_obj_get_int(args[1]);
+  int col = mp_obj_get_int(args[4]);
+
+  uint8_t* font = tm12x6_font;
+  int clen = ((int*)font)[5];
+  int cheight = ((int*)font)[6];
+  int cwidth = ((int*)font)[7];
+
+  int xoffs = 0;
+  while(*data != 0)
+  {
+	  for(int i = 0; i < cheight; i++)
+	  {
+		  for(int j = 0; j < cwidth; j++) // TODO: only works for single byte rows
+		  {
+			  if( font[4*8+(int)(*data)*clen+i] & (1<<(7-j)))
+			  {
+				  if(col == 0)
+					  PXb(x0+j+xoffs,y0+i);
+				  else
+					  PXw(x0+j+xoffs,y0+i);
+			  }
+		  }
+	  }
+	  data++;
+	  xoffs += cwidth;
+
+	  if(*data == '\n')
+	  {
+		  xoffs = 0;
+		  y0 += cheight;
+		  data++;
+	  }
+  }
+
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_string_obj, 5, 5, gfx_string);
+
+
+STATIC mp_obj_t gfx_flush(mp_uint_t n_args, const mp_obj_t *args) {
+	freedomgfxDraw();
+	return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_flush_obj, 0, 1, gfx_flush);
+
+
+// callback system
+
+STATIC uint8_t button_init_done = 0;
+STATIC mp_obj_t button_callbacks[1+BADGE_BUTTONS];
+
+static void gfx_input_poll(uint32_t btn)
+{
+	if(button_init_done && btn > 0 && btn <= BADGE_BUTTONS)
+	{
+		if(button_callbacks[btn] != mp_const_none)
+			mp_sched_schedule(button_callbacks[btn], mp_obj_new_bool(0));
+	}
+}
+
+STATIC mp_obj_t gfx_input_init(mp_uint_t n_args, const mp_obj_t *args) {
+	for(size_t i = 0; i <= BADGE_BUTTONS; i++){
+		button_callbacks[i] = mp_const_none;
+	}
+	button_init_done = 1;
+	return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_input_init_obj, 0, 1, gfx_input_init);
+
+STATIC mp_obj_t gfx_input_attach(mp_uint_t n_args, const mp_obj_t *args) {
+  uint8_t button = mp_obj_get_int(args[0]);
+  button_callbacks[button] = args[1];
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_input_attach_obj, 2, 2, gfx_input_attach);
+
+
+// Module globals
+
+STATIC const mp_rom_map_elem_t freedomgfx_module_globals_table[] = {
+    {MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_freedomgfx)},
+
+    {MP_OBJ_NEW_QSTR(MP_QSTR_init), (mp_obj_t)&gfx_init_obj},
+    {MP_OBJ_NEW_QSTR(MP_QSTR_deinit), (mp_obj_t)&gfx_deinit_obj},
+    {MP_OBJ_NEW_QSTR(MP_QSTR_poll), (mp_obj_t)&gfx_poll_obj},
+
+    {MP_OBJ_NEW_QSTR(MP_QSTR_line), (mp_obj_t)&gfx_line_obj},
+    {MP_OBJ_NEW_QSTR(MP_QSTR_area), (mp_obj_t)&gfx_area_obj},
+	{MP_OBJ_NEW_QSTR(MP_QSTR_string), (mp_obj_t)&gfx_string_obj},
+	{MP_OBJ_NEW_QSTR(MP_QSTR_flush), (mp_obj_t)&gfx_flush_obj},
+
+	{MP_OBJ_NEW_QSTR(MP_QSTR_input_init), (mp_obj_t)&gfx_input_init_obj},
+	{MP_OBJ_NEW_QSTR(MP_QSTR_input_attach), (mp_obj_t)&gfx_input_attach_obj},
+
+    {MP_OBJ_NEW_QSTR(MP_QSTR_BLACK), MP_OBJ_NEW_SMALL_INT(0)},
+    {MP_OBJ_NEW_QSTR(MP_QSTR_WHITE), MP_OBJ_NEW_SMALL_INT(1)},
```

Comment:
If you use 0xff as white, you can write the color directly to the buffer and avoid the if statement.